### PR TITLE
feat(cb2-5690): implement the hgv & trl brakes section

### PR DESCRIPTION
**Pull request by tahayazici**
## Amend PSV Brakes section

- Implement the brakes section for HGVs and TRLs.
- Make the autocomplete component resizable.
- Update the reference data service and store to support the brakes data.
- Add mock data for brakes.
- Create a prefix directive and integrate it into our external autocomplete tool.

[CB2-5690](https://dvsa.atlassian.net/browse/CB2-5690)

### CODE REVIEW

File: src/app/forms/custom-sections/psv-brakes/psv-brakes.component.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  get brakesForm(): FormGroup {
+    return this.form.get('brakes') as FormGroup;
+  }
+
+  get booleanOptions(): MultiOptions {
+    return [{ value: true, label: 'Yes' }, { value: false, label: 'No' }];
+  }
+
+  get retarderOptions(): MultiOptions {
```
Comment (aaron-downing) [inline src/app/forms/custom-sections/psv-brakes/psv-brakes.component.ts:85]: Do we not have an enum for retarder options? is it not something we can pass through that enum to options util func?

File: src/app/forms/templates/general/hgv-trl-brakes.template.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import { ValidatorNames } from '@forms/models/validators.enum';
 import { FormNode, FormNodeTypes } from '../../services/dynamic-form.types';
 
-export const TrlBrakes: FormNode = {
-  name: 'trlBrakesSection',
+export const HgvAndTrlBrakesTemplate: FormNode = {
```
Comment (aaron-downing) [inline src/app/forms/templates/general/hgv-trl-brakes.template.ts:4]: HGVs don't have any brake values, only PSV and TRL:  https://dvsa.atlassian.net/wiki/spaces/HVT/pages/18286839/View+Technical+Record

**Review by aaron-downing — COMMENTED** [review]
If we're completing the amend brakes for TRL in this PR, can we make sure that if there is a ticket for amending TRL, that that ticket is also progressed and tested at the same time, or if a ticket doesn't exist, that we amend the current ticket so that the scope includes TRL and not just PSV

File: src/app/forms/templates/psv/psv-brakes.template.ts [hunk 3]
```diff
@@ -1,49 +1,67 @@
...
       name: 'brakes',
       type: FormNodeTypes.GROUP,
       children: [
         {
           name: 'brakeCode',
-          label: 'Brake Code',
+          label: 'Brake code',
           value: '',
           type: FormNodeTypes.CONTROL,
+          validators: [{ name: ValidatorNames.Required }]
```
Comment (aaron-downing) [inline src/app/forms/templates/psv/psv-brakes.template.ts:18]: Should be 6 char max length

File: src/app/forms/templates/general/hgv-trl-brakes.template.ts [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
+  name: 'hgvAndTrlBrakesSection',
   label: 'Brakes',
```
Comment (aaron-downing) [inline src/app/forms/templates/general/hgv-trl-brakes.template.ts:6]: brakeActuator and leverLength need NUMBER edit type adding to the form group.

Comment (aaron-downing) [reply src/app/forms/templates/general/hgv-trl-brakes.template.ts:6]: antiLockBrakingSystem also needs to be required

File: src/app/forms/custom-sections/psv-brakes/psv-brakes.component.ts (hunk 1 above)
Comment (tahayazici) [reply src/app/forms/custom-sections/psv-brakes/psv-brakes.component.ts:85]: I found the `Retarder` `enum` and I am using that instead now.

Thanks for the heads up.

File: src/app/forms/templates/general/hgv-trl-brakes.template.ts (hunk 2 above)
Comment (tahayazici) [reply src/app/forms/templates/general/hgv-trl-brakes.template.ts:4]: I checked [this ticket](https://dvsa.atlassian.net/browse/CB2-5701) and got confused by its first line of description. You are right. The brakes section for HGVs is not removed.

File: src/app/forms/templates/psv/psv-brakes.template.ts (hunk 3 above)
Comment (tahayazici) [reply src/app/forms/templates/psv/psv-brakes.template.ts:18]: That was omitted intentionally, as we are not allowing for free text, rather we are using the autocomplete field. Just for you, I still included the validator 🙂

**Comment by tahayazici** [comment]
@aaron-downing the work for the TRL brakes section is already done here [CB2-5701](https://dvsa.atlassian.net/browse/CB2-5701). I just included some clean ups for it here as they are related.

File: src/app/forms/templates/general/hgv-trl-brakes.template.ts (hunk 4 above)
Comment (aaron-downing) [reply src/app/forms/templates/general/hgv-trl-brakes.template.ts]: It seems that brakeActuator and leverLength don't have validation on them though even on develop which they should 👀

Comment (aaron-downing) [reply src/app/forms/templates/general/hgv-trl-brakes.template.ts]: Are they making sure that it sends the numbers as a number type and not a string?

File: src/app/forms/custom-sections/psv-brakes/psv-brakes.component.html [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      <th scope="col" class="govuk-table__header"></th>
+      <th scope="col" class="govuk-table__header">No wheels locked</th>
+      <th scope="col" class="govuk-table__header">Up to and including half wheels locked</th>
+    </tr>
+  </thead>
+
+  <tbody class="govuk-table__body">
+    <tr class="govuk-table__row">
+      <th scope="row" class="govuk-table__header">Service</th>
+      <td class="govuk-table__cell">{{ ((vehicleTechRecord.grossLadenWeight || 0) * 16) / 100 }}</td>
```
Comment (aaron-downing) [inline src/app/forms/custom-sections/psv-brakes/psv-brakes.component.html:70]: Numbers need to be rounded to nearest whole number

**Review by aaron-downing — COMMENTED** [review]
Did service design say to use dropdown instead of radio button for retarder selection? Also is it possible to add the 'Brake force requirements' title below the retarder selections

image

**Review by aaron-downing — COMMENTED** [review]
Did we say the first 3 digits of the brake code was out of scope as it's missing at the moment? I realise we need to calculate it from the laden weight

File: src/app/forms/templates/general/hgv-trl-brakes.template.ts (hunk 4 above)
Comment (tahayazici) [reply src/app/forms/templates/general/hgv-trl-brakes.template.ts]: Yep

File: src/app/forms/custom-sections/psv-brakes/psv-brakes.component.html (hunk 5 above)
Comment (tahayazici) [reply src/app/forms/custom-sections/psv-brakes/psv-brakes.component.html:70]: Done

**Comment by tahayazici** [comment]
@aaron-downing I do not recall why I made them dropdowns instead 🙃

They are changed to radio button groups now.

File: src/app/features/tech-record/components/vehicle-technical-record/vehicle-technical-record.component.ts [hunk 6]
```diff
@@ -82,9 +82,17 @@ export class VehicleTechnicalRecordComponent implements OnInit, AfterViewInit {
 
   handleFormState() {
     if (this.isEditing) {
+      const isPsv = this.vehicleTechRecord?.techRecord.find(record => record.statusCode === StatusCodes.CURRENT)?.vehicleType === VehicleTypes.PSV;
```
Comment (tom-bjss) [inline src/app/features/tech-record/components/vehicle-technical-record/vehicle-technical-record.component.ts:85]: We've got this code in a few places now, we should look to move vehicle type checking into the service, as its consumed in atleast 3 places. Not one for now but for everyone to be aware of.

**Review by tom-bjss — APPROVED** [review]

File: src/app/features/tech-record/components/vehicle-technical-record/vehicle-technical-record.component.ts (hunk 6 above)
Comment (tahayazici) [reply src/app/features/tech-record/components/vehicle-technical-record/vehicle-technical-record.component.ts]: Sure, thanks for the heads up.

**Review by kpearmanbjss — DISMISSED** [review]

**Review by tom-bjss — DISMISSED** [review]

**Comment by HazelWilderspin** [comment]
The json model from the tsv looks like this `  {
    "resourceType": "tyres",
    "resourceKey": "A**",
    "code": "A**",
    "dateTimeStamp": "08-MAY-97",
    "description": "3 AXLE SPLIT ( AXLE 1 / AXLES 1+2+3 )"
  },
  {
    "resourceType": "tyres",
    "resourceKey": "F**",
    "code": "F**",
    "dateTimeStamp": "08-MAY-97",
    "description": "3 AXLE SPLIT ( AXLES 1 + 2 / AXLES 2 + 3)"
  },
  {
    "resourceType": "tyres",
    "resourceKey": "6**",
    "code": "6**",
    "dateTimeStamp": "08-MAY-97",
    "description": "2 AXLE SPLIT ( AXLE 1/ AXLE 1 +2 )"
  },
  {
    "resourceType": "tyres",
    "resourceKey": "**6",
    "code": "**6",
    "dateTimeStamp": "08-MAY-97",
    "description": "AXLES 2 + 3"
  },`

Is that going to mess up stuff for you?

**Comment by tahayazici** [comment]
@aaron-downing the prefix is now added. The system did not contain such a component, so implementing it from scratch took some time. It was tricky making it integrate with the external autocomplete tool that we are using.

**Comment by HazelWilderspin** [comment]
Your brakes component doesn't trigger a state update like the others do

**Comment by tahayazici** [comment]
@HazelWilderspin The form is sending the update request with the right data.

**Comment by tahayazici** [comment]
@HazelWilderspin About the HTTP error, I am looking into it. The ticket denotes those fields, so they should be allowed.

**Comment by tahayazici** [comment]
> @HazelWilderspin About the HTTP error, I am looking into it. The ticket denotes those fields, so they should be allowed.

The UI is done and works as expected. This is a bug with the API which I'm looking at with Tom.

**Review by kpearmanbjss — APPROVED** [review]

**Review by kpearmanbjss — APPROVED** [review]

**Review by naathanbrown — APPROVED** [review]

**Pull request merged**